test/modules/layers/test_codebook.py

```diff
@@ -184,3 +186,21 @@ def test_init_embedding_smaller_encoded(self):
                     for enc in encoded_small_flat
                 ]
             ), "embedding initialized from encoder output incorrectly"
+
+    def test_codebook_restart(self):
+        # First init and diversify embedding
+        encoded_flat, _ = self.vq._init_embedding_and_preprocess(self.encoded)
+        # Use only embedding vector at index = 1 and force restarts.
+        # Slightly modify encoded_flat to make sure vectors restart to something new
+        encoded_flat_noise = encoded_flat + torch.randn_like(encoded_flat)
+        codebook_indices_low_usage = torch.ones(encoded_flat.shape[0], dtype=torch.long)
+        self.vq._ema_update_embedding(encoded_flat_noise, codebook_indices_low_usage)
+
+        # Check if embedding contains restarts
+        for i, emb in enumerate(self.vq.embedding):
+            # We used only emb vector with index = 1, so skip because it was not restarted
+            if i == 1:
+                continue
```

Comment:
anyway to check this didnt lead to restart?
for my understanding, can u explain how  the nnext line checks for restart?

Reply:
good point, I should check that this remained the same. the next line compares every embedding vector to every encoded vector (the all part) and sees if any of them are true (which means there is a single match and there was a restart). couldn't use assert_expected because of this logic